Mergin/version_viewer_dialog.py

```diff
@@ -0,0 +1,660 @@
+            self.stackedWidget.setCurrentIndex(0)
+            self.tabWidget.setCurrentIndex(0)
+            self.tabWidget.setTabEnabled(0, True)
+            layers = self.collect_layers(self.toggle_layers_action.isChecked())
+            self.update_canvas(layers)
+        else:
+            self.toolbar.setEnabled(False)
+            self.stackedWidget.setCurrentIndex(1)
+            self.label_info.setText("No visual changes")
+            self.tabWidget.setCurrentIndex(1)
+            self.tabWidget.setTabEnabled(0, False)
+
+    def collect_layers(self, checked: bool):
+        if checked:
+            layers = iface.mapCanvas().layers()
+
+            # Filter only "Background" type
+            whitelist_backgound_layer_types = [QgsRasterLayer, QgsVectorTileLayer, QgsTiledSceneLayer]
+            layers = [layer for layer in layers if type(layer) in whitelist_backgound_layer_types]
+        else:
+            layers = []
+
+        if self.current_diff:
+            layers.insert(0, self.current_diff)
+
+        return layers
+
+    def diff_layer_changed(self, index: int):
+        if index > len(self.diff_layers) or index < 0:
+            return
+
+        self.map_canvas.setLayers([])
+        self.attribute_table.clearSelection()
+
+        self.current_diff = self.diff_layers[index]
+
+        self.layer_cache = QgsVectorLayerCache(self.current_diff, 1000)
+        self.layer_cache.setCacheGeometry(False)
+
+        self.table_model = QgsAttributeTableModel(self.layer_cache)
+        self.table_model.setRequest(QgsFeatureRequest().setFlags(QgsFeatureRequest.NoGeometry).setLimit(100))
+
+        self.filter_model = QgsAttributeTableFilterModel(self.map_canvas, self.table_model)
+
+        self.layer_cache.setParent(self.table_model)
+
+        self.attribute_table.setModel(self.filter_model)
+        self.table_model.loadLayer()
+
+        config = self.current_diff.attributeTableConfig()
+        self.filter_model.setAttributeTableConfig(config)
+        self.attribute_table.setAttributeTableConfig(config)
+
+        layers = self.collect_layers(self.toggle_layers_action.isChecked())
+        self.update_canvas(layers)
+
+    def add_current_to_project(self):
+        if self.current_diff:
+            QgsProject.instance().addMapLayer(self.current_diff)
+
+    def add_all_to_project(self):
+        for layer in self.diff_layers:
+            QgsProject.instance().addMapLayer(layer)
+
+    def zoom_full(self):
+        if self.current_diff:
+            layerExtent = self.current_diff.extent()
+            # transform extent
+            layerExtent = self.map_canvas.mapSettings().layerExtentToOutputExtent(self.current_diff, layerExtent)
+
+            self.map_canvas.setExtent(layerExtent)
+            self.map_canvas.refresh()
+
+    def zoom_selected(self):
+        if self.current_diff:
+            self.map_canvas.zoomToSelected([self.current_diff])
+            self.map_canvas.refresh()
```

Comment:
A little pedantic, but neither the list nor the refresh should be needed here.
```suggestion
            self.map_canvas.zoomToSelected(self.current_diff)
```